src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/RestrictedMethodFinder.java

```diff
@@ -0,0 +1,150 @@
+
+    private final Map<MethodRef, Boolean> CACHE = new HashMap<>();
+    private final ClassResolver classesToScan;
+    private final ClassResolver systemClassResolver;
+
+    private RestrictedMethodFinder(ClassResolver classesToScan, ClassResolver systemClassResolver) {
+        this.classesToScan = classesToScan;
+        this.systemClassResolver = systemClassResolver;
+    }
+
+    public static RestrictedMethodFinder create(ClassResolver classesToScan, ClassResolver systemClassResolver) throws JNativeScanFatalError, IOException {
+        return new RestrictedMethodFinder(classesToScan, systemClassResolver);
+    }
+
+    public Map<ScannedModule, Map<ClassDesc, List<RestrictedUse>>> findAll() throws JNativeScanFatalError {
+        Map<ScannedModule, Map<ClassDesc, List<RestrictedUse>>> restrictedMethods = new HashMap<>();
+        classesToScan.forEach((_, info) -> {
+            ClassModel classModel = info.model();
+            List<RestrictedUse> perClass = new ArrayList<>();
+            classModel.methods().forEach(methodModel -> {
+                if (methodModel.flags().has(AccessFlag.NATIVE)) {
+                    perClass.add(new NativeMethodDecl(MethodRef.ofModel(methodModel)));
+                } else {
+                    Set<MethodRef> perMethod = new HashSet<>();
+                    methodModel.code()
+                        .ifPresent(code -> {
+                            code.forEach(e -> {
+                                switch (e) {
+                                    case InvokeInstruction invoke -> {
+                                        if (isRestrictedMethod(invoke.method())) {
+                                            perMethod.add(MethodRef.ofMethodRefEntry(invoke.method()));
+                                        }
+                                    }
+                                    default -> {
+                                    }
+                                }
+                            });
+                        });
+                    if (!perMethod.isEmpty()) {
+                        perClass.add(new RestrictedMethodRefs(MethodRef.ofModel(methodModel),
+                                Set.copyOf(perMethod)));
+                    }
+                }
+            });
+            if (!perClass.isEmpty()) {
+                ScannedModule scannedModule = new ScannedModule(info.jarPath(), info.moduleName());
+                restrictedMethods.computeIfAbsent(scannedModule, _ -> new HashMap<>())
+                        .put(classModel.thisClass().asSymbol(), perClass);
+            }
+        });
+        return restrictedMethods;
+    }
+
+    private boolean isRestrictedMethod(MemberRefEntry method) throws JNativeScanFatalError {
+        return switch (method) {
+            case MethodRefEntry mre ->
+                    isRestrictedMethod(mre.owner().asSymbol(), mre.name().stringValue(), mre.typeSymbol());
```

Comment:
Do we really need these two identical calls to `isRestrictedMethod` ? Note that both `MethodRefEntry` and `InterfaceMethodEntry` are subclasses of `MemberRefEntry`, so perhaps the other `isRestrictedMethod` can just take `MemberRefEntry`?

Or maybe we can have different factories in `MethodRef` one that takes `InterfaceMethodEntry` and another that takes `MemberRefEntry`.

Reply:
I can do that, but I think this will always be a bit awkward since these types don't have a common super type that exposes the needed information.

Reply:
> these types don't have a common super type that exposes the needed information

No wait, they actually do :) That's just `MemberRefEntry`.

Reply:
I was able to eliminate this method entirely, now the caller just creates a `MethodRef` and passes that to the other `isRestrictedMethod` overload.